### PR TITLE
cloud: switch to AWS SDK v2

### LINKED ISSUE #116320: Transition from AWS SDK V1 to V2
CRDB currently uses the [AWS SDK V1 for go](https://github.com/aws/aws-sdk-go) which does not support some features, for example [multi-region access points for s3](https://docs.aws.amazon.com/sdkref/latest/guide/feature-s3-mrap.html). 

CRDB already uses [AWS SDK V2](https://github.com/aws/aws-sdk-go-v2) for some DMS integrations and testing.

**Describe the solution you'd like**
We standardize S3 and KMS access through V2 of the SDK and remove V1 of the SDK. We would need to migrate [s3_storage.go](https://github.com/cockroachdb/cockroach/blob/28457a0ac0464f58fe617728139bd1134dbe75ca/pkg/cloud/amazon/s3_storage.go) and [aws_kms.go](https://github.com/cockroachdb/cockroach/blob/28457a0ac0464f58fe617728139bd1134dbe75ca/pkg/cloud/amazon/aws_kms.go) plus tests over to the new SDK.

Jira issue: CRDB-34577

### PR DESCRIPTION
**Pull request by msbutler**
Fixes: https://github.com/cockroachdb/cockroach/issues/116320
Release note(build change): changes the AWS SDK version used
for interactions with external storage from v1 to v2.

### CODE REVIEW

**Comment by msbutler** [comment]
Clean  cloud nightly run [here](https://teamcity.cockroachdb.com/agent/3013546).

**Comment by msbutler** [comment]
unrelated flake

**Comment by dt** [comment]
the pts test failure is certainly unrelated but I'm kicking off a new CI just since we want to see green here.

File: pkg/cloud/amazon/aws_kms.go [hunk 1]
```diff
@@ -128,38 +129,41 @@ func MakeAWSKMS(ctx context.Context, uri string, env cloud.KMSEnv) (cloud.KMS, e
 	}
 
 	region := kmsURIParams.region
-	awsConfig := &aws.Config{
-		Credentials: credentials.NewStaticCredentials(kmsURIParams.accessKey,
-			kmsURIParams.secret, kmsURIParams.tempToken),
+	if region == "" {
+		// TODO(adityamaru): Maybe use the KeyID to get the region, similar to how
+		// we infer the region from the bucket for s3_storage.
+		return nil, errors.New("could not find the aws kms region")
```
Comment (dt) [inline pkg/cloud/amazon/aws_kms.go:135]: is this going to be a user-visible behavior regression?

**Review by dt — APPROVED** [review]

File: pkg/cloud/amazon/aws_kms.go (hunk 1 above)
Comment (msbutler) [reply pkg/cloud/amazon/aws_kms.go]: Not a documented one. https://www.cockroachlabs.com/docs/stable/take-and-restore-encrypted-backups#aws-kms-uri-format. 

But i can do a little more digging once this merges.

**Review by stevendanna — APPROVED** [review]
If our tests are passing, then I think we should default to merging this.  In the last run at this, we were seeing problems with implicit auth using instance credentials (i.e. no env var, just creds via the metadata service).  But...I think that would have showed up in any of the AWS roachtests, so either your cleanup fixed it or some bug fix we pulled in fixed it.

**Comment by msbutler** [comment]
TFTRs!

bors r+

File: pkg/cloud/amazon/aws_kms.go (hunk 1 above)
Comment (msbutler) [reply pkg/cloud/amazon/aws_kms.go]: verified that this is not a regression. If the user does not specify a region, they hit this error (before this pr):
https://github.com/msbutler/cockroach/blob/butler-kms-no-region/pkg/cloud/amazon/aws_kms.go#L222

Comment (dt) [reply pkg/cloud/amazon/aws_kms.go]: Thanks for checking!

**Pull request merged**